Fix JSON Editor toggling

matborowczyk opened the pull request:
# Description

Fix switching between JSON editor and Input Form converting values to string when it should not.
Converting to string will now only happen if the dict is invalid

closes #6177 

https://github.com/user-attachments/assets/0206934b-e10d-4ed8-8e25-e538e150670a


# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Code is clear and sufficiently documented
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

LukasStordeur commented on src/UI/Components/ServiceInstanceForm/Helpers/createFormState.spec.ts:
Descriptions of the tests are the same.
Aren't the createEditFormState and createDuplicateFormState basically the same? I don't mind asserting them too, but might feel a bit redundant. 
It does also happen that we get from the backend the dict as either an object, or stringified object (with escaped characters if my memory serves me well.)

LukasStordeur commented on src/UI/Components/ServiceInstanceForm/Helpers/createFormState.spec.ts:
```diff
@@ -20,7 +88,21 @@ test("GIVEN createFormState WHEN passed a dict field with a default value THEN c
+  });
+});
+
+test("Given creteFormState WHEN passed nested fields state THEN creates formState correctly", () => {
```
```suggestion
test("Given createFormState WHEN passed nested fields state THEN creates formState correctly", () => {
```

LukasStordeur commented on src/UI/Components/ServiceInstanceForm/Helpers/createFormState.ts:
```diff
@@ -93,9 +101,22 @@ export const createEditFormState = (
-          acc[curr.name] = curr.type.includes("dict")
-            ? stringifyDict(originalAttributes?.[curr.name])
-            : cloneDeep(originalAttributes?.[curr.name]);
+          if (curr.type === "int?" || curr.type === "float?") {
```
I'm confused, if we have a Text field, or TextList, how is it we have the possibility that we are dealing with integers or floats? is it purely to cover the empty state `""`?

LukasStordeur commented on src/UI/Components/ServiceInstanceForm/Helpers/createFormState.ts:
```diff
@@ -93,9 +101,22 @@ export const createEditFormState = (
+          if (curr.type === "int?" || curr.type === "float?") {
+            const value = originalAttributes?.[curr.name];
+
+            //empty string assertion is for converting input form to JSON Editor
```
This is not clear to me. 
The JSON editor is relying on the JSON-Schema. 
We cannot assign null like that there when the field is empty. 
I feel like I'm missing the actual problem.

LukasStordeur commented on src/UI/Components/ServiceInstanceForm/Helpers/createFormState.ts:
```diff
@@ -173,9 +194,22 @@ export const createDuplicateFormState = (
-          acc[curr.name] = curr.type.includes("dict")
-            ? stringifyDict(originalAttributes?.[curr.name])
-            : cloneDeep(originalAttributes?.[curr.name]);
+          if (curr.type === "int?" || curr.type === "float?") {
```
This feels redundant, and not optimized. 
The formState is a complex function, adding even more complexity will make it hard to read.

LukasStordeur reviewed (commented):
Hmm, Not entirely certain we are approaching the problem correctly.

LukasStordeur commented on src/UI/Components/ServiceInstanceForm/Components/FieldInput.tsx:
```diff
@@ -223,9 +223,18 @@ export const FieldInput: React.FC<Props> = ({
-          handleInputChange={(value, _event) =>
-            getUpdate(makePath(path, field.name), value)
-          }
+          handleInputChange={(value, _event) => {
```
We are doing this manipulation in other places too, checking if something is parseable to JSON. I'd suggest the following and reuse it where we are doing this try catch :

```ts
const handleInputChange = (value, _event) => {
  if (field.type.includes("dict")) {
    getUpdate(makePath(path, field.name), tryParseJSON(value));
  } else {
    getUpdate(makePath(path, field.name), value);
  }
};

const tryParseJSON = (value) => {
  try {
    return JSON.parse(value);
  } catch (_error) {
    // If the value is not a valid JSON string, return the original value
    return value;
  }
};
```

matborowczyk commented:
I addressed all remarks and according to our call.

Improved test cases to better reflect the issue, unified test as code added was extracted and reused so there is no room for only on formstate creator to success/fail in this scenario

I checked other inputs and no other place is parsingJSON or work with the input in similar fashion where that parsing could be implemented but reused it in conversion.

inmantaci commented:
Processing this pull request

inmantaci commented:
Merged into branches master, iso8 in 81cddda53a0ce1fa7161ec390cad384f95b3ac9a
